text/0000-cfg-path-version.md

```diff
@@ -0,0 +1,928 @@
+    v.push(1);
+}
+```
+
+Note that you won't be able to make use of `#[cfg(version(..))]` for these 
+particular features since they were introduced before this RFC's features
+get stabilized. This means that you can't for example add `version(1.28)`
+to your code and expect Rust 1.28 compilers to enable the code.
+However, there will be features in the future to use this mechanism on.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `#[cfg(version(<semver>))]`
+
+To the `cfg` attribute, a `version` flag is added.
+This flag has the following grammar (where `\d` is any digit in `0` to `9`):
+
+```rust
+flag : "version" "(" semver ")" ;
+semver : digits ("." digits ("." digits)?)? ;
+digits : \d+ ;
+```
+
+[caret requirements]: https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#caret-requirements
+
+If and only if a Rust compiler considers itself to have a version which is
+greater or equal to the version in the `semver` string will the
+`#[cfg(version(<semver>)]` flag be considered active.
+Greater or equal is defined in terms of [caret requirements].
+
+## `#[cfg(accessible($path))]`
+
+To the `cfg` attribute, a `accessible` flag is added.
+This flag requires that a `path` fragment be specified in it inside parenthesis
+but not inside a string literal. The `$path` must start with leading `::`
+and may not refer to any parts of the own crate (e.g. with `::crate::foo`,
+`::self::foo`, or `::super::foo` if such paths are legal).
+This restriction exists to ensure that the user does not try to
+conditionally compile against parts of their own crate because that crate
+has not been compiled when the `accessible` flag is checked on an item.
+
+If and only if the path referred to by `$path` does exist and is public
+will the `#[cfg(accessible($path))]` flag be considered active.
+In checking whether the path exists or not, the compiler will consider
+feature gated items to exist if the gate has been enabled.
+
+If a path refers to an item inside an inherent implementation,
+the path will be considered to exist if any configuration of generic
+parameters can lead to the item. To check whether an item exists for
+an implementation with a specific sequence of concrete types applied to
+a type constructor, it is possible to use the `::foo::bar::<T>::item` syntax.
+
+It is also possible to refer to fields of `struct`s, `enum`s, and `unions`.
+Assuming that we have the following definitions in the `foobar` crate:
+
+```rust
+pub struct Person { pub ssn: SSN, age: u16 }
+
+pub enum Shape<Unit> {
+    Triangle { pub sides: [Unit; 3] },
+    ...
+}
+
+pub union MaybeUninit<T> { uninit: (), pub value: T }
+```
+
+We can then refer to them like so:
+
+```rust
+#[cfg(all(
+    accessible(::foobar::Person::ssn),
+    accessible(::foobar::Shape::Triangle::sides),
+    accessible(::foobar::Shape::MaybeUninit::value)
+))]
+fn do_stuff() {
+    ...
+}
+```
```

Comment:
What happens if there are overlapping fields and inherent items in a type? e.g. if the above `Person` struct was extended in the future with an inherent method `pub fn ssn(&self) -> Result<ValidatedSSN, ValidationError>` would there be any way to distinguish between this method and the `ssn` field?

Reply:
I don't think you could distinguish. Similarly, if you had a macro named `foo` and a function named `foo` you couldn't distinguish between those either since `accessible` considers all namespaces. If you wanted this level of accuracy then, as a potential future work, `accessible(field, ::foobar::Person::ssn)` and `accessible(fn, ::foobar::Person::ssn)` could be possible.